Change representation for fingerprints

GambolingPangolin opened the pull request:
This changes how we represent fingerprints to `newtype Fingerprint = Fingerprint { unFingerprint :: Word32 }`.  This has a few benefits.  First, `Fingerprint` has `Serialize` etc. instances and downstream structures use these instances, so there is no risk of the encoding drifting from one place to another.  (The binary encodings for fingerprints in the extended keys section disagreed with that from the psbt section.)   Second, I added IsString/Read/Show instances that treat `Fingerprint` as having an 8-digit hex string representation instead of a decimal number.

GambolingPangolin committed:
Harmonize byte order

This makes the serialization of fingerprints and key indices for PSBTs
match the serialization used in the extended key module

wraithm commented on src/Haskoin/Transaction/Partial.hs:
```diff
@@ -679,15 +679,15 @@ instance Serialize PSBTHDPath where
         S.putByteString bs
       where
-        bs = S.runPut $ S.putWord32le fp >> mapM_ S.putWord32le kis
+        bs = S.runPut $ S.putWord32be fp >> mapM_ S.putWord32le kis
```
What about the other word32le there?

Is there mixed byteorder?

Also, I'm surprised the test suite doesn't catch this!!

Would you mind checking if the psbt test suite is missing vectors?

GambolingPangolin replied:
We might also need to change that.  `DerivPathI` uses big-endian byte order for key indices.  I'll add test vectors as needed and make sure that what we do conforms with the BIP 174 spec.  The BIP has a lot of test vectors.

GambolingPangolin replied:
The situation is messy:
- BIP 32 specifies big-endian for key indices
- BIP 174 specifies little-endian for HD path components

Also, I added a test case covering HD paths.

wraithm commented on test/Haskoin/Transaction/PartialSpec.hs:
```diff
@@ -137,6 +203,7 @@ vec6Test = do
         (Key 0x0f (fromJust $ decodeHex "010203040506070809"))
         (fromJust $ decodeHex "0102030405060708090a0b0c0d0e0f")
 
+
```
nit: unnecessary whitespace

wraithm commented on test/Haskoin/Transaction/PartialSpec.hs:
```diff
@@ -121,6 +119,74 @@ vec5Test = do
+                    ]
+        }
+    hardIndex = (+ 2^31)
+
```
nit: unnecessary whitespace

GambolingPangolin committed:
Use a newtype for Fingerprint

This also changes the read/show instances to use hex strings and adds an
IsString instance.

wraithm commented on haskoin-core.cabal:
```diff
@@ -86,7 +86,7 @@ library
       Haskoin.Util.Arbitrary.Transaction
       Haskoin.Util.Arbitrary.Util
   other-modules:
-      Paths_haskoin_core
```
Why remove this?

wraithm replied:
Oh, it's automatic! hmm

GambolingPangolin replied:
This must have been a side effect of _adding_ an `other-modules` field to `package.yaml`.  Do we use the paths module for anything?  The project builds without it.  Maybe we should leave it off.

wraithm replied:
Yeah, let's maybe leave it off. The paths modules cause all kinds of problems with test rebuilding and stuff... Would you mind actually explicitly ignoring it (like we do in our repo)?

GambolingPangolin replied:
Sure!

wraithm commented on src/Haskoin/Keys/Extended/Internal.hs:
```diff
@@ -0,0 +1,61 @@
+     where
+       decodeError  = error "Fingerprint literal: Unable to decode"
+       hexError = error "Fingerprint literal: Invalid hex"
+
```
nit: 2 newlines

wraithm commented on CHANGELOG.md:
```diff
@@ -4,6 +4,11 @@ All notable changes to this project will be documented in this file.
+## 0.21.0
+### Changed
+- Use a newtype for Fingerprint, which uses an 8 digit hex string for IsString,
+  Read, & Show.  This fixes inconsistent (de)serialization across the package.
```
It might be good to mention the Paths_haskoin_core thing.

@jprupp, what do you think about this? I don't imagine that people are depending on this module, but you can tell me otherwise.

wraithm replied:
We can also take the conservative route and still expose it.

GambolingPangolin replied:
`Paths_haskoin_core` was not part of the public API.  It was under `other-modules` and now it's unlisted altogether.

wraithm replied:
Ah, okay! Nvm then. Resolving.
